horovod/run/elastic/registration.py

```diff
@@ -0,0 +1,149 @@
+                # such, we need to update the state to FAILURE, and we don't want two threads coming from the same
+                # worker at the barrier.
+                #
+                # In order to ensure that the new failing thread can record results in cases of total job failure,
+                # we also need to block this thread by waiting on the barrier. This requires us to reset the barrier,
+                # as otherwise this worker will be double-counted (once for the READY thread and once for FAILURE),
+                # which would cause the barrier to complete too early.
+                logging.info('key exists, reset barrier: {}[{}] = {}'.format(host, slot, state))
+                self._barrier.reset()
+            logging.info('record state: {}[{}] = {}'.format(host, slot, state))
+            self._states[key] = state
+            self._workers[state].add(key)
+            rendezvous_id = self._rendezvous_id
+
+        rendezvous_id = self._wait(key, state, rendezvous_id)
+        return rendezvous_id
+
+    def _wait(self, key, state, rendezvous_id):
+        while True:
+            try:
+                self._barrier.wait()
+                return rendezvous_id
+            except threading.BrokenBarrierError:
+                if self._barrier.broken:
+                    # Timeout or other non-recoverable error, so exit
+                    raise
+
+                # Barrier has been reset
+                with self._lock:
+                    # Check to make sure the reset was not caused by a change of state for this key
+                    rendezvous_id = self._rendezvous_id
+                    saved_state = self._states.get(key, state)
+                    if saved_state != state:
+                        # This worker changed its state, so do not attempt to wait again to avoid double-counting
+                        raise RuntimeError('State {} overridden by {}'.format(state, saved_state))
+
+    def _action(self):
+        self._on_workers_recorded()
+
+    def _on_workers_recorded(self):
+        logging.info('all {} workers recorded'.format(self.size()))
+
+        # Check for success state, if any process succeeded, shutdown all other processes
+        if self.count(SUCCESS) > 0:
+            logging.info('success count == {} -> stop running'.format(self.count(SUCCESS)))
+            self._driver.stop()
+            return
+
+        # Check that all processes failed, indicating that processing should stop
+        if self.count(FAILURE) == self._size:
+            logging.error('failure count == {} -> stop running'.format(self._size))
+            self._driver.stop()
+            return
+
+        # Check for failures, and add them to the blacklisted hosts list
+        failures = self.get(FAILURE)
+        for host, slot in failures:
+            self._discovered_hosts.blacklist(host)
+
+        # If there are no active hosts that aren't blacklisted, treat this as job failure
```

Comment:
this comment is really hard to understand, it is double-negated, please rephrase

Reply:
Reworded.